pygmt/clib/session.py

```diff
@@ -1430,14 +1435,25 @@ def virtualfile_from_data(self, check_kind=None, data=None, x=None, y=None, z=No
         if kind in ("file", "grid"):
             _data = (data,)
         elif kind == "vectors":
-            _data = (x, y, z)
+            _data = [np.atleast_1d(x), np.atleast_1d(y)]
+            if z is not None:
+                _data.append(np.atleast_1d(z))
+            if extra_arrays:
+                _data.extend(extra_arrays)
         elif kind == "matrix":  # turn 2D arrays into list of vectors
             try:
                 # pandas.DataFrame and xarray.Dataset types
                 _data = [array for _, array in data.items()]
             except AttributeError:
-                # Python lists, tuples, and numpy ndarray types
-                _data = np.atleast_2d(np.asanyarray(data).T)
+                try:
+                    # Just use virtualfile_from_matrix for 2D
+                    # numpy.ndarray which are not datetime (M) types
+                    assert data.ndim == 2 and not data.dtype.kind == "M"
```

Comment:
> assert data.ndim == 2 and not data.dtype.kind == "M"

The [`GMT_Put_Matrix()`](https://docs.generic-mapping-tools.org/dev/api.html#gmt-put-matrix) function only supports a few [numeric data types](https://docs.generic-mapping-tools.org/dev/api.html#tbl-types), but [numpy.dtype.kind](https://numpy.org/doc/stable/reference/generated/numpy.dtype.kind.html) can have values other than `M`, for example, `data.dtype.kind == "O"` will pass the assert statement but is not supported by `virtualfile_from_matrix`. Perhaps check `data.dtype.kind in 'iuf'`?

Reply:
Ok, I'll try `data.dtype.kind in 'iuf'` and see if it works on the test suite.